### PR TITLE
Add profiling to compyle

### PR DESCRIPTION
**Pull request by adityapb**

### CODE REVIEW

File: compyle/array.py [hunk 1]
```diff
@@ -456,11 +460,14 @@ def get_allocator(queue):
...
 def sort_by_keys(ary_list, out_list=None, key_bits=None,
-                 backend=None, use_radix_sort=False):
+                 backend=None, use_radix_sort=False,
+                 profile_id="sort_by_keys"):
     # FIXME: Need to use returned values, cuda backend uses
     # thrust that will internally allocate a new array for storing
     # the sorted data so out_list will not have the sorted arrays
     # first arg of ary_list is the key
+    result = None
+    start = time.time()
```
Comment (prabhuramachandran) [inline compyle/array.py:470]: This block looks like something perfect for a context manager.  So a `with profile('name'): ` may be a way to facilitate this?

File: compyle/config.py [hunk 2]
```diff
@@ -14,7 +14,7 @@ def __init__(self):
         self._use_cuda = None
         self._use_double = None
         self._omp_schedule = None
-        self._profile = None
+        self._profile_kernels = None
```
Comment (prabhuramachandran) [inline compyle/config.py:17]: Why is this changed from `profile`?

File: compyle/cuda.py [hunk 3]
```diff
@@ -35,7 +35,11 @@ def set_context():
 
 class SourceModule(_SourceModule):
     def __getattr__(self, name):
-        return self.get_function(name)
+        def kernel(*args, **kwargs):
+            f = self.get_function(name)
+            return f(*args, **kwargs)
+        kernel.function_name = name
```
Comment (prabhuramachandran) [inline compyle/cuda.py:41]: Can't you set `kernel.__name__` do you need another attribute?

Comment (adityapb) [reply compyle/cuda.py:41]: The pyopencl.Program uses function_name so I wanted to keep it the same for cuda kernels as well

File: compyle/array.py [hunk 4]
```diff
@@ -455,6 +459,7 @@ def get_allocator(queue):
     return allocator
 
 
+@profile_method
```
Comment (prabhuramachandran) [inline compyle/array.py:462]: Seems that most times the public API only uses this so it seems that a shorter name would be better, perhaps just `profile` is better?

File: compyle/opencl.py [hunk 5]
```diff
@@ -2,14 +2,12 @@
 """
 from __future__ import print_function
 import pyopencl as cl
-from collections import defaultdict
-from operator import itemgetter
 
 from .config import get_config
+from .profile import profile_kernel, record_profile, named_profile
```
Comment (prabhuramachandran) [inline compyle/opencl.py:7]: Is `named_profile` used anywhere or is record profile used?

File: compyle/profile.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+import time
+from contextlib import contextmanager
+from collections import defaultdict
+from .config import get_config
+
+
+_profile_info = defaultdict(lambda: {'calls': 0, 'time': 0})
+
+
+def record_profile(name, time):
```
Comment (prabhuramachandran) [inline compyle/profile.py:13]: Seems like this is now an implementation detail so this can be a private function,`_record_profile`.

File: compyle/profile.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    global _profile_info
+    _profile_info[name]['time'] += time
+    _profile_info[name]['calls'] += 1
+
+
+@contextmanager
+def profile(name):
```
Comment (prabhuramachandran) [inline compyle/profile.py:20]: This is the only problem, what do we call this?  Perhaps `profile` can be an object which has an `__enter__/__exit__` and a `__call__` so it can work as either a decorator or a context manager?  Otherwise, if you expect this to be used less, perhaps `profile_ctx`?

File: compyle/profile.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    start = time.time()
+    yield start
+    end = time.time()
+    record_profile(name, end - start)
+
+
+def profile_method(method):
```
Comment (prabhuramachandran) [inline compyle/profile.py:27]: This can be `profile`.

File: compyle/profile.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    def wrapper(*args, **kwargs):
+        self = args[0] if len(args) else None
+        with profile(getattr(self, "name", method.__name__)):
+            return method(*args, **kwargs)
+    return wrapper
+
+
+def print_profile():
+    global _profile_info
```
Comment (prabhuramachandran) [inline compyle/profile.py:36]: Is this global declaration needed?

File: compyle/profile.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+    for kernel, data in profile_data:
+        print("{:<40} {:<10} {:<10}".format(
+            kernel,
+            data['calls'],
+            data['time']))
+        tot_time += data['time']
+    print("Total profiled time: %g secs" % tot_time)
+
+
+def profile_kernel(kernel, name, backend=None):
```
Comment (prabhuramachandran) [inline compyle/profile.py:54]: Some docstring with an example or additional documentation would be useful, there are now many public functions and it isn't clear what is to be used.  I have myself forgotten!

File: compyle/profile.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+    if get_config().profile:
+        wgi = getattr(kernel, 'get_work_group_info', None)
+        if wgi is not None:
+            _profile_knl.get_work_group_info = wgi
+        return _profile_knl
+    else:
+        return kernel
+
+
+def named_profile(name, backend=None):
```
Comment (prabhuramachandran) [inline compyle/profile.py:84]: This doesn't seem to be used anywhere?  Or is it only used in pysph?  Perhaps some tests are in order in that case as there seem to be no tests for any of this.  Sorry I am hassling you.

File: compyle/profile.py (hunk 9 above)
Comment (adityapb) [reply compyle/profile.py:58]: I think this will be needed to record profile, right? Can't think of a way around it

File: compyle/profile.py (hunk 11 above)
Comment (adityapb) [reply compyle/profile.py:109]: This is only used in pysph. I have added some tests for these

File: compyle/opencl.py (hunk 5 above)
Comment (adityapb) [reply compyle/opencl.py]: In pysph `named_profile` is imported from `compyle.opencl` so I added it here. I can change all imports in pysph so that they import from `compyle.profile` in a later PR

Comment (prabhuramachandran) [reply compyle/opencl.py]: OK.

File: compyle/profile.py (hunk 9 above)
Comment (prabhuramachandran) [reply compyle/profile.py:58]: The print is just reading the _profile_info and not writing to it so it should not require the global, Python should look in the global scope since the name is not in the function scope.

File: compyle/profile.py (hunk 11 above)
Comment (prabhuramachandran) [reply compyle/profile.py:109]: Great, thanks.

**Review by prabhuramachandran — APPROVED** [review]

**Comment by prabhuramachandran** [comment]
Looks good to me, PySPH depends on master so please send a corresponding PR to pysph when possible.
